lib/SILOptimizer/Mandatory/OwnershipModelEliminator.cpp

```diff
@@ -57,16 +57,35 @@ namespace {
 /// consistent non-ossa vs ossa code rather than an intermediate state.
 struct OwnershipModelEliminatorVisitor
     : SILInstructionVisitor<OwnershipModelEliminatorVisitor, bool> {
-  SILBuilder &builder;
   SmallVector<SILInstruction *, 8> trackingList;
   SmallBlotSetVector<SILInstruction *, 8> instructionsToSimplify;
+
+  /// Points at either a user passed in SILBuilderContext or points at
+  /// builderCtxStorage.
+  SILBuilderContext builderCtx;
+
   SILOpenedArchetypesTracker openedArchetypesTracker;
 
-  OwnershipModelEliminatorVisitor(SILBuilder &newBuilder)
-      : builder(newBuilder),
-        openedArchetypesTracker(&newBuilder.getFunction()) {
-    newBuilder.setTrackingList(&trackingList);
-    newBuilder.setOpenedArchetypesTracker(&openedArchetypesTracker);
+  /// Construct an OME visitor for eliminating ownership from \p fn.
+  OwnershipModelEliminatorVisitor(SILFunction &fn)
+      : trackingList(), instructionsToSimplify(),
+        builderCtx(fn.getModule(), &trackingList),
+        openedArchetypesTracker(&fn) {
+    builderCtx.setOpenedArchetypesTracker(&openedArchetypesTracker);
+  }
+
+  /// A "syntactic" high level function that combines our insertPt with a
+  /// builder ctx.
+  ///
+  /// Since this is syntactic and we assume that our caller is passing in a
+  /// lambda that if we inline will be eliminated, we mark this function always
+  /// inline.
+  template <typename ResultTy>
+  ResultTy LLVM_ATTRIBUTE_ALWAYS_INLINE
+  withBuilder(SILInstruction *insertPt,
+              llvm::function_ref<ResultTy(SILBuilder &, SILLocation)> visitor) {
+    SILBuilderWithScope builder(insertPt, builderCtx);
+    return visitor(builder, insertPt->getLoc());
```

Comment:
It's too bad this is so awkward. Ideally, building a single instruction should be able to naturally use function chaining:

`SILBuilder(inst, builderCtx).emitSomethingElse(operands)`

Reply:
Yes. But today it is SILBuilderWithScope.